src/Control/Parallel/Class.purs

```diff
@@ -138,6 +147,49 @@ instance monadRaceWriterT :: (Monoid w, MonadRace m) => MonadRace (WriterT w m)
   stall = WriterT stall
   race (WriterT w1) (WriterT w2) = WriterT (race w1 w2)
 
+-- | The `MonadFork` class abstracts over monads which support some notion of
+-- | being forked from the current "thread".
+-- |
+-- | `fork` and `cancelWith` should obey the following law:
+-- | - ``fork m >>= cancel e = fork (m `cancelWith` Canceler (const (pure true))) >>= cancel e``
+class Monad m <= MonadFork m where
+  fork :: forall a. m a -> m (Canceler m)
+  cancelWith :: forall a. m a -> Canceler m -> m a
+
+instance monadForkExceptT :: MonadFork m => MonadFork (ExceptT r m) where
+  fork (ExceptT ma) = ExceptT (Right <<< mapCanceler lift <$> fork ma)
+  cancelWith (ExceptT ma) c =
+    ExceptT (cancelWith ma (mapCanceler (\(ExceptT mb) -> mb >>= either (const (pure false)) pure) c))
+
+instance monadForkMaybeT :: MonadFork m => MonadFork (MaybeT m) where
+  fork (MaybeT ma) = MaybeT (Just <<< mapCanceler lift <$> fork ma)
+  cancelWith (MaybeT ma) c =
+    MaybeT (cancelWith ma (mapCanceler (\(MaybeT mb) -> mb >>= maybe (pure false) pure) c))
+
+instance monadForkReaderT:: MonadFork m => MonadFork (ReaderT r m) where
+  fork (ReaderT ma) =
+    ReaderT \r -> mapCanceler lift <$> fork (ma r)
+  cancelWith (ReaderT ma) c =
+    ReaderT \r -> cancelWith (ma r) (mapCanceler (\(ReaderT mb) -> mb r) c)
+
+-- | A canceler for a forked computation.
+newtype Canceler m = Canceler (CancelReason -> m Boolean)
+
+instance semigroupCanceler :: Apply m => Semigroup (Canceler m) where
+  append (Canceler f1) (Canceler f2) = Canceler (\e -> (||) <$> f1 e <*> f2 e)
+
+instance monoidCanceler :: Applicative m => Monoid (Canceler m) where
+  mempty = Canceler (const (pure true))
+
+-- | The reason for canceling a forked computation.
+newtype CancelReason = CancelReason String
```

Comment:
I think it would be nice to have the class parameterized by the reason type (`class MonadFork r m | m -> r`, obviously requiring functional dependencies). Though I don't know how useful the "reason" actually is. I don't know of any usages with Aff that have required observing the error.